iib/workers/tasks/opm_operations.py

```diff
@@ -918,3 +922,259 @@ def verify_operator_exists(
             is_package_in_index = True
             log.info("operator package %s found in index_db %s", operator_package, index_db_path)
     return is_package_in_index, index_db_path
+
+
+@retry(
+    before_sleep=before_sleep_log(log, logging.WARNING),
+    reraise=True,
+    retry=retry_if_exception_type(IIBError),
+    stop=stop_after_attempt(2),
+)
+def opm_index_add(
+    base_dir: str,
+    bundles: List[str],
+    binary_image: str,
+    from_index: Optional[str] = None,
+    graph_update_mode: Optional[str] = None,
+    overwrite_from_index_token: Optional[str] = None,
+    overwrite_csv: bool = False,
+    container_tool: Optional[str] = None,
+) -> None:
+    """
+    Add the input bundles to an operator index.
+
+    This only produces the index.Dockerfile file and does not build the container image.
+
+    :param str base_dir: the base directory to generate the database and index.Dockerfile in.
+    :param list bundles: a list of strings representing the pull specifications of the bundles to
+        add to the index image being built.
+    :param str binary_image: the pull specification of the container image where the opm binary
+        gets copied from. This should point to a digest or stable tag.
+    :param str from_index: the pull specification of the container image containing the index that
+        the index image build will be based from.
+    :param str graph_update_mode: Graph update mode that defines how channel graphs are updated
+        in the index.
+    :param str overwrite_from_index_token: the token used for overwriting the input
+        ``from_index`` image. This is required to use ``overwrite_from_index``.
+        The format of the token must be in the format "user:password".
+    :param bool overwrite_csv: a boolean determining if a bundle will be replaced if the CSV
+        already exists.
+    :param str container_tool: the container tool to be used to operate on the index image
+    :raises IIBError: if the ``opm index add`` command fails.
+    """
+    # The bundles are not resolved since these are stable tags, and references
+    # to a bundle image using a digest fails when using the opm command.
+
+    from iib.workers.tasks.utils import run_cmd, set_registry_token
```

Comment:
Is there a reason for doing the same import multiple times at this place (opm_index_rm, derpecate_bundles too)?

Reply:
I have mostly moved these functions but I also believe these were imported withi the function to avoid circular dependencies

Reply:
we should really clean up these imports and fix circular dependencies, but I believe that should be in scope of another PR